src/avram/attribute.cr

```diff
@@ -59,11 +66,13 @@ class Avram::Attribute(T)
     errors.empty?
   end
 
-  def value=(@value)
-    @changed = true
+  def changed?(**arguments)
```

Comment:
The issue with using named args is that you could also pass random arguments and they’d be silently ignored. 

It also makes the documentation a bit less clear since it isn’t clear what args it accepts. 

I think what we could do here instead is make 4 different changed? Methods with different args using method overloading. So one with no args, one with to, one with from, and another with both. 

The type for to/from could also be restricted to T so that you can only pass in values that make sense

Does this make sense?

Reply:
> I think what we could do here instead is make 4 different changed? Methods with different args using method overloading. So one with no args, one with to, one with from, and another with both.

That is exactly how I implemented it at first:

```cr
def changed?
  @value != @original_value
end

def changed?(from : T)
  changed? && from == @value
end

def changed?(to : T)
  changed? && to == @value
end

def changed?(from : T, to : T)
  changed?(from: from) && changed?(to: to)
end
```

But that does not work because `changed?(to : T)` will overwrite `changed?(from : T)` as they have the same argument type. In this scenario, we would have to ditch `changed?(to : T)` because that is the least interesting option. On the other hand, that would make the implementation seem incomplete.

Another solution I was thinking of would be:

```cr
def changed_from?(value : T)
  changed? && value == @value
end

def changed_to?(value : T)
  changed? && value == @value
end
```

But then what doe we do with the from/to situation? We could keep:

```cr
def changed?(from : T, to : T)
  changed_from?(from) && changed_to?(to)
end
```
Although, I think that's making for a confusing api since `changed?(from : T)` does not exist. Therefore, the only option would be to:

```cr
def changed_from_to?(from : T, to : T)
  changed_from?(from) && changed_to?(to)
end
```

There is no ideal solution, unless I am missing something. :smile:

Reply:
Right, just remembered I had two other solutions in mind, but they are not very sexy:


```cr
def changed?(value : T, to : Bool)
  changed? && value == (to ? @value : @original_value)
end

# changed?("value") => testing from
# changed?("value", true) => testing to
```
:mask: 
```cr
def changed?(way : Symbol, value : T)
  changed? && value == (way == :to ? @value : @original_value)
end

# changed?(:from, "value")
# changed?(:to, "value")

# And consequently:
# changed?(:from, "value", :to, "other_value")
```
Please forgive the uglyness. :blush:

Reply:
Ah yes I see what you mean. There is a workaround that I used for save operations. I created a `Nothing` class that represents an unused argument in a save operation. So argument defaults are Nothing.new. That way someone can pass “nil” and we know they meant to pass that since if they passed no argument it’d be Nothing. 

https://github.com/luckyframework/avram/blob/f533f088c24d44c53baa48d8bf6d0c823c93fd36/src/avram/needy_initializer_and_save_methods.cr

https://github.com/luckyframework/avram/blob/f533f088c24d44c53baa48d8bf6d0c823c93fd36/src/avram/needy_initializer_and_save_methods.cr#L77

https://github.com/luckyframework/avram/blob/f533f088c24d44c53baa48d8bf6d0c823c93fd36/src/avram/needy_initializer_and_save_methods.cr#L92

So I think we could have one method with two args  for “from” and “to”. The arts would default to Nothing.new so that you could check for is_a?(Nothing) like in the link I posted. Does that make sense?

Reply:
Yes, perfect. Works like a charm! Thanks for pointing me in the right direction. :)